File: solutions/r/diamond/1/diamond.R
diamond <- function(letter) {
  letters_up_to_input <- LETTERS[1:match(letter, LETTERS)]
  diamond_letters <- c(
    letters_up_to_input,
    rev(letters_up_to_input[-length(letters_up_to_input)])
  )

  max_index <- match(letter, LETTERS)
  width <- 2 * max_index - 1

  make_row <- function(current_letter) {
    index <- match(current_letter, LETTERS)

    outer_spaces <- max_index - index
    inner_spaces <- 2 * index - 3

    if (current_letter == "A") {
      row <- paste0(
        strrep(" ", outer_spaces),
        current_letter,
        strrep(" ", outer_spaces)
      )
    } else {
      row <- paste0(
        strrep(" ", outer_spaces),
        current_letter,
        strrep(" ", inner_spaces),
        current_letter,
        strrep(" ", outer_spaces)
      )
    }

    row
  }

  rows <- vapply(diamond_letters, make_row, character(1))

  paste(rows, collapse = "\n")
}

File: solutions/cpp/diamond/1/diamond.cpp
#include <cstdint>
#include "diamond.h"
using namespace std;
auto diamond::rows(char c) -> vector<string>
{
  auto size = (c - 'A') * 2 + 1;
  auto mid = size / 2;
  auto diamond = vector(size, string(size, ' '));
  for(uint8_t i = 0; i <= mid; i++) {
    diamond[i][mid + i] = 'A' + i;
    diamond[i][mid - i] = 'A' + i;
    diamond[size - i - 1][mid + i] = 'A' + i;
    diamond[size - i - 1][mid - i] = 'A' + i;
  }
  return diamond;
}
